### PR TITLE
[rllib] Add option for RNN state and value estimates to span episodes

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

The use case here is an infinite horizon MDP, in which case you still need to get rewards for the episode by setting a horizon, but don't want a finite horizon for optimization purposes.

## Related issue number

Closes https://github.com/ray-project/ray/issues/4425

### CODE REVIEW

File: python/ray/rllib/agents/agent.py [hunk 1]
```diff
@@ -66,8 +66,13 @@
...
     # Discount factor of the MDP
     "gamma": 0.99,
-    # Number of steps after which the episode is forced to terminate
+    # Number of steps after which the episode is forced to terminate. Defaults
+    # to `env.spec.max_episode_steps` (if present) for Gym envs.
     "horizon": None,
+    # Calculate rewards but don't reset the environment when the horizon is
+    # hit. This allows value estimation and RNN state to span across logical
+    # episodes denoted by horizon. This only has an effect if horizon != inf.
+    "soft_horizon": False,
```
Comment (ericl) [inline python/ray/rllib/agents/agent.py]: Better name?

Comment (hartikainen) [reply python/ray/rllib/agents/agent.py]: Maybe change `horizon -> max_episode_length` and `soft_horizon -> horizon`?

**Review by hartikainen — APPROVED** [review]
The horizon config naming could be made a bit clearer, but otherwise looks good.

**Comment by ericl** [comment]
jenkins retest this please

File: python/ray/rllib/agents/agent.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/agents/agent.py]: Yeah that's a bit clearer. It might not be worth renaming horizon thou, we'd have to deprecate it.

**Pull request merged**